### PR TITLE
[babel 8] Use an identifier for `TSTypeParameter.name`

### LINKED ISSUE #12825: parser(ts): TSTypeParameter name should be an Identifier
## Feature Request

- [ ] I would like to work on this feature!

#### This change can/should be considered as a breaking change.

**Is your feature request related to a problem?**

Make TSTypeParameter name property an Identifier.
There is no easy way to traverse over this field and this is also not aligned with other uses of generic/template parameters/arguments 

TSTypeReference and TSTypeParameter has simmilar use -> both can be used in generic/template parameters

```ts
f<F>()

TSTypeReference
  - typeName: Identifier
```

```ts
function f<T>() {}

TSTypeParameter
  - name: string
```

```ts
function f< /* x */ T = 2>() {}
function f<T>() {}
function f<T = 2>() {}
```

**Describe the solution you'd like**

`TSTypeReference` node should have property name of type `Identifier` instead of `string`

**Describe alternatives you've considered**.

additionally we could align also name of property `typeName` and `name` (I'm not sure about this)

**Documentation, Adoption, Migration Strategy**

### Additional info

prettier integration issue https://github.com/JamesHenry/typescript-estree/issues/109 (yes old one i forgot to repost it here)

### PR DESCRIPTION
**Pull request by fedeci**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #12825  
| Patch: Bug Fix?          |
| Major: Breaking Change?  | Yes, under `BABEL_8_BREAKING` and `BABEL_TYPES_8_BREAKING`
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

We parse the entire identifier if `BABEL_8_BREAKING` otherwise we parse just his name.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12829"></a>

### CODE REVIEW

File: packages/babel-parser/src/types.js [hunk 1]
```diff
@@ -956,7 +956,12 @@ export type TsTypeParameterDeclaration = TypeParameterDeclarationBase & {
 };
 
 export type TypeParameterBase = NodeBase & {
+  // TODO(Babel-8): remove name property
+  /**
+   * @deprecated use `typeName` instead
+   */
   name: string,
+  typeName: Identifier,
```
Comment (fedeci) [inline packages/babel-parser/src/types.js:964]: We can also handle this as a union type assigning it under the `BABEL_8_BREAKING` flag.
```suggestion
  name: string | Identifier,
```

**Review by nicolo-ribaudo — CHANGES_REQUESTED** [review]
Let's just do this for Babel 8 (`BABEL_8_BREAKING`).

This PR as-is is problematic because it introduces two different source of truth in the AST, and modifying `typeName` in a plugin doesn't have any effect on the generated code.

**Comment by fedeci** [comment]
Those ~30 errors are all related to `babel-parser` fixtures that need to be updated. Is there a way to avoid duplicating them?

File: packages/babel-parser/src/types.js [hunk 2]
```diff
@@ -965,7 +965,8 @@ export type TsTypeParameterDeclaration = TypeParameterDeclarationBase & {
 };
 
 export type TypeParameterBase = NodeBase & {
-  name: string,
+  // TODO(Babel-8): remove string type support
+  name: string | Identifier,
 };
 
 export type TypeParameter = TypeParameterBase & {
```
Comment (fedeci) [inline packages/babel-parser/src/types.js:972]: `TypeParameter` also depends on `TypeParameterBase`, this must be changed too, right?

**Comment by nicolo-ribaudo** [comment]
> Those ~30 errors are all related to babel-parser fixtures that need to be updated. Is there a way to avoid duplicating them?

You need to duplicate them, rename the copy as `...-babel-7`, and add `BABEL_8_BREAKING: true` and `BABEL_8_BREAKING: false` to the test options.

**Comment by fedeci** [comment]
It's huge

**Review by nicolo-ribaudo — APPROVED** [review]

**Review by nicolo-ribaudo — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
I re-approved because I'm not sure about what changed since my last review, but I reviewed it again.

**Review by existentialism — APPROVED** [review]

**Comment by fedeci** [comment]
Just wait to merge this that I have to update `TypeParameter` also.

**Review by nicolo-ribaudo — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
> Since Flow parses `TypeParameters.name` as a string we don't have to change that behaviour.

We might still decide that it's better to align our Flow implementation with our TS implementation, rather than aligning it with the Flow AST. However, that's a separate discussion and let's keep this PR only for TS.

**Pull request merged**